homeassistant/components/nibe_heatpump/config_flow.py

```diff
@@ -0,0 +1,131 @@
+"""Config flow for Nibe Heat Pump integration."""
+from __future__ import annotations
+
+import errno
+import logging
+from typing import Any
+
+from nibe.connection.nibegw import NibeGW
+from nibe.exceptions import CoilReadException, CoilWriteException
+from nibe.heatpump import HeatPump, Model
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_IP_ADDRESS, CONF_MODEL
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import config_validation as cv
+from homeassistant.util.network import is_ipv4_address
+
+from .const import (
+    CONF_CONNECTION_TYPE,
+    CONF_CONNECTION_TYPE_NIBEGW,
+    CONF_LISTENING_PORT,
+    CONF_REMOTE_READ_PORT,
+    CONF_REMOTE_WRITE_PORT,
+    CONF_WORD_SWAP,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_MODEL): vol.In([e.name for e in Model]),
+        vol.Required(CONF_IP_ADDRESS): str,
+        vol.Required(CONF_LISTENING_PORT): cv.port,
+        vol.Required(CONF_REMOTE_READ_PORT): cv.port,
+        vol.Required(CONF_REMOTE_WRITE_PORT): cv.port,
+    }
+)
+
+
+class FieldError(Exception):
+    """Field with invalid data."""
+
+    def __init__(self, field: str, error: str) -> None:
+        """Set up error."""
+        super().__init__(f"Error `{error}` in `{field}`")
+        self.field = field
+        self.error = error
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect."""
+
+    if not is_ipv4_address(data[CONF_IP_ADDRESS]):
+        raise FieldError(CONF_IP_ADDRESS, "address")
+
+    heatpump = HeatPump(Model[data[CONF_MODEL]])
+    heatpump.initialize()
+
+    connection = NibeGW(
+        heatpump,
+        data[CONF_IP_ADDRESS],
+        data[CONF_REMOTE_READ_PORT],
+        data[CONF_REMOTE_WRITE_PORT],
+        listening_port=data[CONF_LISTENING_PORT],
+    )
+
+    try:
+        await connection.start()
+    except OSError as exception:
+        if exception.errno == errno.EADDRINUSE:
+            raise FieldError("listening_port", "address_in_use") from exception
+        raise
+
+    try:
+        coil = heatpump.get_coil_by_address(48852)
```

Comment:
Should `48852` be encased in a constant?

Reply:
I can grab it by name here instead, then it is a bit clearer.